dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/merge/MergeProcessor.java

```diff
@@ -0,0 +1,773 @@
+    public static OpenAPIProvider mergeDocuments(List<OpenAPIProvider> documents) {
+
+        MergeProcessor mergeProcessor = new MergeProcessor();
+        
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            Tr.event(mergeProcessor, tc, "Beginning merge of OpenAPI documents for " + documents);
+        }
+
+        for (OpenAPIProvider document : documents) {
+            mergeProcessor.add(document);
+        }
+
+        mergeProcessor.process();
+
+        OpenAPIProvider mergedDoc = mergeProcessor.getMergedDocument();
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            try {
+                Tr.event(mergeProcessor, tc, "Merged document: ", OpenApiSerializer.serialize(mergedDoc.getModel(), Format.JSON));
+            } catch (IOException e) {
+                Tr.event(mergeProcessor, tc, "Unable to trace merged document", e);
+            }
+        }
+
+        return mergedDoc;
+
+    }
+
+    /**
+     * The providers to merge
+     */
+    private List<OpenAPIProvider> providers = new ArrayList<>();
+
+    /**
+     * Models which have been processed to remove clashes
+     */
+    private List<OpenAPI> processedModels = new ArrayList<>();
+
+    /**
+     * This list of providers used to create the merged document
+     */
+    private List<OpenAPIProvider> includedProviders = new ArrayList<>();
+
+    /**
+     * The path names in use for models processed so far
+     */
+    private Map<OpenAPIProvider, Set<String>> pathNames = new HashMap<>();
+
+    /**
+     * The unique names and their values which are in use for models processed so far, grouped by the type of name
+     */
+    private Map<NameType, Map<String, Object>> namesInUse = new HashMap<>();
+
+    private List<String> mergeProblems = new ArrayList<>();
+
+    private void add(OpenAPIProvider document) {
+        providers.add(document);
+    }
+
+    private void process() {
+
+        List<InProgressModel> inProgressModels = new ArrayList<>();
+
+        // First loop which just does all the steps necessary to determine whether there are any fatal merge conflicts
+        // which result in removing some of the requested merge providers.
+        // This ensures that when we come to rename things to avoid non-fatal merge conflicts, we aren't trying to avoid
+        // clashing with models that have been removed.
+        for (OpenAPIProvider provider : providers) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Starting path processing for " + provider);
+            }
+
+            // Make a deep copy, to avoid modifying the input model
+            OpenAPI model = (OpenAPI) ModelCopy.copy(provider.getModel());
+
+            RenameHolder renameHolder = new RenameHolder();
+
+            // Add context root to the front of paths if possible
+            prependPaths(model, provider.getApplicationPath(), renameHolder);
```

Comment:
instead of creating a new RenameHolder and passing it in, wouldn't it be cleaner to have prependPaths create and return the RenameHolder?

Reply:
Probably not. `prependPaths` deals with moving the context root from servers to paths, there's no reason it should create the `renameHolder`. `renameHolder` is also passed as a parameter in the same way to other methods further down.